### ISSUE TITLE
probably connection leak after redis cluster add a slave node

### ISSUE THREAD

**Issue by polarisGitHub** [redisson/redisson#5922]
### Redis version

5.0.12

### Redisson version

3.15.5/3.16.4

### Redisson configuration

```ymal
clusterServersConfig:
  idleConnectionTimeout: 10000
  connectTimeout: 10000
  timeout: 1000
  retryAttempts: 3
  retryInterval: 1500
  failedSlaveReconnectionInterval: 3000
  failedSlaveCheckInterval: 60000
  password: 
  subscriptionsPerConnection: 5
  clientName: null
  loadBalancer: !<org.redisson.connection.balancer.RoundRobinLoadBalancer> {}
  subscriptionConnectionMinimumIdleSize: 1
  subscriptionConnectionPoolSize: 50
  slaveConnectionMinimumIdleSize: 24
  slaveConnectionPoolSize: 180
  masterConnectionMinimumIdleSize: 24
  masterConnectionPoolSize: 180
  readMode: "MASTER"
  subscriptionMode: "MASTER"
  nodeAddresses:
    - ""
  scanInterval: 3000
  pingConnectionInterval: 30000
  keepAlive: false
  tcpNoDelay: true
threads: 16
nettyThreads: 32
codec: !<org.redisson.codec.JsonJacksonCodec> {}
transportMode: "EPOLL"
```

### What is the Expected behavior?

use synchronized when add slave or release connection in addSlaveEntry method

### What is the Actual behavior?

After add a slave node to a healthy redis cluster, the number of connection of new slave node become particularly high. 

When I checked redisson log and read  source code of **ClusterConnectionManager**,I suspected there was a concurrency issue in **addRemoveSlaves**

## 3.15.5 source code
 **entry.addSlave(uri, configEndpointHostName);** Do not check if the node was added, so we can see a node being added multiple times.
![slave](https://github.com/redisson/redisson/assets/9389651/a09ad2eb-4630-4c38-a321-43d3fd9d877b)

```java
private CompletableFuture<Set<RedisURI>> addRemoveSlaves(MasterSlaveEntry entry, ClusterPartition currentPart, ClusterPartition newPart) {

            // we can  add node multiple times here
            CompletableFuture<Void> slaveUpFuture = entry.addSlave(uri, configEndpointHostName);
            CompletableFuture<Void> f = slaveUpFuture.thenAccept(res -> {
                currentPart.addSlaveAddress(uri);
                log.info("slave: {} added for master {} and slot ranges: {}",
                        currentPart.getMasterAddress(), uri, currentPart.getSlotRanges());
                entry.excludeMasterFromSlaves(uri);
            });
            futures.add(f);
        }
    }
```

## 3.16.4 fixed once
This problem was fixed once  in version 3.16.4  [3871 Fixed - recovered slave shouldn't be added again in Redis Cluster mode
](https://github.com/redisson/redisson/issues/3871)
But I don't think it has been completely solved, even in the 3.30.0 ,the **ClusterConnectionManager** refactored 

In 3.16.4, before adding the slave node, the code first checks whether the entry has already added the node.However, if multiple schedulers detect node changes at the same time, or the link initialization is slow, it may still be added multiple times. The root cause of this problem is a concurrency issue

```java
for (RedisURI uri : addedSlaves) {
            ClientConnectionsEntry slaveEntry = entry.getEntry(uri);

            // check fisrt, but Cannot fight against concurrency
            if (slaveEntry != null) {
                currentPart.addSlaveAddress(uri);
                entry.slaveUp(uri, FreezeReason.MANAGER);
                log.info("slave: {} added for slot ranges: {}", uri, currentPart.getSlotRanges());
                continue;
            }

            RFuture<Void> future = entry.addSlave(uri);
}
```

## 3.30.0
in latest version ,3.30. I found that when **addSlaveEntry** actually doing the add, the code checked again, When the client has been added, it will be refused to be added to pools, but the established connection will not be released again.

```java
public CompletableFuture<Void> addSlave(RedisClient client) {
        noPubSubSlaves.set(false);
        CompletableFuture<InetSocketAddress> addrFuture = client.resolveAddr();
        return addrFuture.thenCompose(res -> {
            ClientConnectionsEntry entry = new ClientConnectionsEntry(client,
                    config.getSlaveConnectionMinimumIdleSize(),
                    config.getSlaveConnectionPoolSize(),
                    connectionManager,
                    NodeType.SLAVE,
                    config);

            CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
            if (!entry.isFreezed()) {
                // connection init first,but do not release in addSlaveEntry when a client has been added in client2Entry
                CompletableFuture<Void> slaveFuture = entry.initConnections(config.getSlaveConnectionMinimumIdleSize());
                CompletableFuture<Void> pubSubFuture = entry.initPubSubConnections(config.getSubscriptionConnectionMinimumIdleSize());
                future = CompletableFuture.allOf(slaveFuture, pubSubFuture);
            }
            return future.thenAccept(r -> {
                addSlaveEntry(entry);
            });
        }).whenComplete((r, ex) -> {
            if (ex != null) {
                client.shutdownAsync();
            }
        });
    }

   private void addSlaveEntry(ClientConnectionsEntry entry) {
        if (client2Entry.get(entry.getClient()) != null) {
           // do not release connections
            return;
        }
        slaveConnectionPool.addEntry(entry);
        slavePubSubConnectionPool.addEntry(entry);
        client2Entry.put(entry.getClient(), entry);
    }
```

**Comment by mrniko** [comment]
Thanks for the details.

In the latest version there is a check for the slave presence executed before addSlave method execution.

![image](https://github.com/redisson/redisson/assets/1104661/7754f7dc-8ad0-42d6-a2c0-1e278ce7f9c7)

**Comment by polarisGitHub** [comment]
I wonder if it is possible to have the following case

When a new slave added, the result obtained by **ClientConnectionsEntry slaveEntry = entry.getEntry(uri);** is null. 

When addSlave is executed slowly, for example due to network problems, link establishment is very slow. It may happen that the first addSlave has not been executed yet, and the next **checkClusterState** starts scheduling execution. At this time, because addSlave has not been executed, the entry data is not updated. **ClientConnectionsEntry slaveEntry = entry.getEntry(uri);** will still be null, so you can execute addSlave again

```java
private synchronized void addSlaveEntry(ClientConnectionsEntry entry) {
        if (client2Entry.get(entry.getClient()) != null) {
            return;
        }
        slaveConnectionPool.addEntry(entry);
        slavePubSubConnectionPool.addEntry(entry);
        // update entry here
        client2Entry.put(entry.getClient(), entry);
    }
```

**Comment by mrniko** [comment]
> It may happen that the first addSlave has not been executed yet, and the next checkClusterState starts scheduling execution. 

It can't be happen since futures are invoked in chain and topology scan is isn't completed in this case.

**Comment by mrniko** [comment]
Can you try the latest version?

**Comment by polarisGitHub** [comment]
The production environment uses an old framework, using redisson 3.16.4, and there are no large-span version updates planned in the near future. Afterwards, new projects can try to use a newer version for observation.

**Comment by mrniko** [comment]
4 issues related to cluster failover handling has been fixed since 3.16.4 version. Please update.
